### DISCUSSION TITLE
Everything seems to be working ... just can't seem to connect to the rtsp stream

### DISCUSSION THREAD

**Discussion by leifhopper** [thirtythreeforty/neolink#313] [Q&A]
First off; thank you for taking the time and effort to create and maintain this project!

I'm trying neolink out on some B500 cameras I have. The documentation and discussions are great to get started and were quite helpful. It seems that I have the dependencies and configuration correct based on the rust info logging to the terminal.

The issue I'm seeing is that regardless of what address or port I bind to, I just can't seem to be able to open the rtsp stream. I've tried connecting to the RTSP stream using VLC on another machine on the network (neolink bound to it's machine network address) as well as tried by adding the stream as a source for ZoneMinder with neolink bound to 127.0.0.1 where ZM and neolink run on the same machine. No dice. If this camera model is out of scope for neolink ... well, it was a nice idea to try ;-) 

Ubuntu Server 20.04.5 LTS
**Firewall disabled**

The config.toml:
(FWIW, I've bound the address to 127.0.0.1, the IP of the machine and as in the config below to no avail. Also tried the port bind on 8554 as well as the config below)
```
bind = "0.0.0.0"
bind_port = 554

[[cameras]]
name = "drive"
username = "123456"
password = "123456"
address = "10.10.10.85:9000"
stream = "mainStream"

```
Bash output:
```
./neolink rtsp --config=config.toml
[2022-09-21T22:11:45Z INFO  neolink] Neolink d75e416c96b9867875bc7252777302e8cec214ab release
[2022-09-21T22:11:45Z INFO  neolink::rtsp] drive: Connecting to camera at Address: 10.10.10.85:9000
[2022-09-21T22:11:45Z INFO  neolink::rtsp] drive: Logging in
[2022-09-21T22:11:45Z INFO  neolink::rtsp] drive: Connected and logged in
[2022-09-21T22:11:45Z INFO  neolink::rtsp] drive: Camera time is already set: 2022-09-21 15:11:46 -7
[2022-09-21T22:11:45Z INFO  neolink::rtsp] drive: Camera reports firmware version v2.0.0.16_20090825
[2022-09-21T22:11:45Z INFO  neolink::rtsp] drive: Starting video stream Sub Stream (Fluent)
```

Anyone have any ideas I might try?

**Comment by leifhopper** [comment]
Sorry, I left out that the rtsp address I tried was 
`rtsp://10.10.10.75:554/drive`

**Reply by QuantumEntangledAndy** [reply]
Ok let's put gstreamer in debug log mode. 

Start neolink with this environmental variable set `GST_DEBUG=3`. You did specify how it was being run (docker or directly) but I hope you can figure out how to add that to your setup. 

Once done use ffprobe on the same machine as neolink.

```bash
ffprobe rtsp://127.0.0.1:554/drive
```

ffprobe will hopefully tell us that it connected and give us the format of the video. 

If not then look at the logs made by neolink as ffprobe was connecting it should give hints as to what is wrong.

**Comment by leifhopper** [comment]
Thanks for the hints @QuantumEntangledAndy 

I'm running neolink directly as part of testing whether neolink would work at all for me. If I can work out this issue, a production environment would be with docker ... but keeping it simple for now.

I exported the env var and set the bind port to 127.0.0.1 to keep as much entropy out of your guided tests as possible.

I tested that `export GST_DEBUG=3` was working by first running neolink as a non-privileged user and I do get additional debug. Code snip below showing that debugging is actually on.

```
0:00:00.004737343  3904 0x55cdbb393c60 ERROR             rtspserver rtsp-server.c:928:gst_rtsp_server_create_socket:<GstRTSPServer@0x55cdbb2dc940> failed to create socket
0:00:00.004753203  3904 0x55cdbb393c60 ERROR             rtspserver rtsp-server.c:1295:gst_rtsp_server_create_source:<GstRTSPServer@0x55cdbb2dc940> failed to create socket
0:00:00.004760909  3904 0x55cdbb393c60 ERROR             rtspserver rtsp-server.c:1341:gst_rtsp_server_attach:<GstRTSPServer@0x55cdbb2dc940> failed to create watch: Error binding to address 127.0.0.1:554: Permission denied
```

Starting neolink with elevated privileges gives the output here:
```
sudo ./neolink rtsp --config=config.toml
[2022-09-22T00:04:07Z INFO  neolink] Neolink d75e416c96b9867875bc7252777302e8cec214ab release
[2022-09-22T00:04:07Z INFO  neolink::rtsp] drive: Connecting to camera at Address: 10.10.10.85:9000
[2022-09-22T00:04:07Z INFO  neolink::rtsp] drive: Logging in
[2022-09-22T00:04:08Z INFO  neolink::rtsp] drive: Connected and logged in
[2022-09-22T00:04:08Z INFO  neolink::rtsp] drive: Camera time is already set: 2022-09-21 17:04:08 -7
[2022-09-22T00:04:08Z INFO  neolink::rtsp] drive: Camera reports firmware version v2.0.0.16_20090825
[2022-09-22T00:04:08Z INFO  neolink::rtsp] drive: Starting video stream Main Stream (Clear)
```

tmuxed a session to test `ffprobe rtsp://127.0.0.1:554/drive` and we get this output

```
[rtsp @ 0x56064a75d300] max delay reached. need to consume packet
[rtsp @ 0x56064a75d300] RTP: missed 232 packets
[h264 @ 0x56064a760c80] error while decoding MB 0 29, bytestream -17
[h264 @ 0x56064a760c80] concealing 14609 DC, 14609 AC, 14609 MV errors in I frame
[rtsp @ 0x56064a75d300] DTS discontinuity in stream 0: packet 21 with DTS 80797, packet 22 with DTS 93026
[rtsp @ 0x56064a75d300] DTS discontinuity in stream 0: packet 24 with DTS 93031, packet 25 with DTS 104432
[rtsp @ 0x56064a75d300] DTS discontinuity in stream 0: packet 26 with DTS 104434, packet 27 with DTS 121941
[rtsp @ 0x56064a75d300] max delay reached. need to consume packet
[rtsp @ 0x56064a75d300] RTP: missed 215 packets
Input #0, rtsp, from 'rtsp://127.0.0.1:554/drive':
  Metadata:
    title           : Session streamed with GStreamer
    comment         : rtsp-server
  Duration: N/A, start: 0.020625, bitrate: N/A
    Stream #0:0: Video: h264 (High), yuv420p(progressive), 2560x1920, 90k tbr, 90k tbn, 180k tbc
    Stream #0:1: Audio: pcm_s16be, 16000 Hz, 2 channels, s16, 512 kb/s

```

I hope I've got here what you're looking for ... and thanks for taking this question up!

**Reply by QuantumEntangledAndy** [reply]
The ffprobe successfully showed the connection is working and a stream is being provided. It was able to detect the format of the video audio and the resolution. 

It did however suggest that the udp buffers are filling up quite fast. Perhaps the video resolution is too high for the default buffer size. This may be why you are not getting a picture in vlc which also has small udp buffer sizes by default. Best way to get around this is usually to force tcp. 

Next test from another machine that you want to view it from run

```bash
ffplay -rtsp_transport tcp -i rtsp://remote_ip:554/drive
```

p.s. Those options might not all be correct as I'm doing it from memory please look them up in the docs if it fails.

**Comment by leifhopper** [comment]
Terribly sorry. I left out that no debug messages come from neolink when ffprobe connects to the rtsp stream address

**Comment by leifhopper** [comment]
Well then. ffplay does indeed play the rtsp stream when the transport is set to tcp.

Confirmed neolink works with the B500 camera as well, then.

I did a bit of digging around buffering in ZoneMinder (my intended solution) and it seems that I can now also read the rtsp stream when setting `Source -> Method: TCP` and `Source -> Options buffer_size=10000000`

I'm one happy camper. Thanks for your help!

**Reply by QuantumEntangledAndy** [reply]
Glad to have helped. You probably lost the env var for debug print when you ran it via sudo. It tends to reset the environment
